File: voicerecognition/src/main/java/guepardoapps/lucahome/voicerecognition/services/VoiceRecognitionService.java
package guepardoapps.lucahome.voicerecognition.services;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.AsyncTask;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;

import java.io.File;
import java.io.IOException;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Locale;

import edu.cmu.pocketsphinx.Assets;
import edu.cmu.pocketsphinx.Hypothesis;
import edu.cmu.pocketsphinx.RecognitionListener;
import edu.cmu.pocketsphinx.SpeechRecognizer;
import edu.cmu.pocketsphinx.SpeechRecognizerSetup;

import guepardoapps.library.openweather.converter.NotificationContentConverter;
import guepardoapps.library.openweather.datatransferobjects.NotificationContentDto;
import guepardoapps.library.openweather.models.ForecastModel;
import guepardoapps.library.openweather.models.WeatherModel;
import guepardoapps.library.openweather.service.OpenWeatherService;

import guepardoapps.lucahome.common.classes.WirelessSocket;
import guepardoapps.lucahome.common.controller.BroadcastController;
import guepardoapps.lucahome.common.utils.Logger;
import guepardoapps.lucahome.voicerecognition.classes.RelationActionClass;
import guepardoapps.lucahome.voicerecognition.utils.IRelationshipHelper;

@SuppressWarnings({"WeakerAccess"})
public class VoiceRecognitionService implements RecognitionListener, ActivityCompat.OnRequestPermissionsResultCallback, IVoiceRecognitionService {
    private static final String Tag = VoiceRecognitionService.class.getSimpleName();

    private static final String KwsSearch = "wakeup";
    private static final String KeyPhrase = "hey luca";

    private static final int PermissionRequestRecordAudio = 345243854;

    private final VoiceRecognitionService Singleton = new VoiceRecognitionService();

    private boolean _isInitialized;
    private boolean _permissionRecordAudioGranted;

    private Context _context;
    private IRelationshipHelper _relationshipHelper;

    private BroadcastController _broadcastController;
    private SpeechRecognizer _speechRecognizer;

    private VoiceRecognitionService() {
    }

    @Override
    public VoiceRecognitionService getInstance() {
        return Singleton;
    }

    @Override
    public InitializeResult Initialize(@NonNull Context context, @NonNull IRelationshipHelper relationshipHelper) {
        if (_isInitialized) {
            Logger.getInstance().Debug(Tag, "Already initialized!");
            return InitializeResult.AlreadyInitialized;
        }

        if (!(context instanceof Activity)) {
            Logger.getInstance().Debug(Tag, "Context is not of type Activity!");
            return InitializeResult.ContextNotOfTypeActivity;
        }

        Logger.getInstance().Debug(Tag, "Initialize");

        _context = context;
        _relationshipHelper = relationshipHelper;
        _broadcastController = new BroadcastController(_context);

        try {
            _permissionRecordAudioGranted = checkRecordAudioPermission();
        } catch (Exception exception) {
            broadcastException(exception);
        }

        if (_permissionRecordAudioGranted) {
            new SetupTask(this).execute();
        }

        _isInitialized = true;
        return InitializeResult.InitializeSuccess;
    }

    @Override
    public void Dispose() {
        Logger.getInstance().Debug(Tag, "Dispose");
        if (_speechRecognizer != null) {
            _speechRecognizer.cancel();
            _speechRecognizer.shutdown();
        }
        _broadcastController = null;
        _isInitialized = false;
    }

    @Override
    public boolean RequestRecordAudioPermission() {
        Logger.getInstance().Verbose(Tag, "RequestRecordAudioPermission");
        if (_permissionRecordAudioGranted) {
            Logger.getInstance().Verbose(Tag, "Already granted!");
            return true;
        }

        try {
            _permissionRecordAudioGranted = checkRecordAudioPermission();
            return _permissionRecordAudioGranted;
        } catch (Exception exception) {
            broadcastException(exception);
            return false;
        }
    }

    @Override
    public void onBeginningOfSpeech() {

    }

    @Override
    public void onEndOfSpeech() {
        if (!_speechRecognizer.getSearchName().equals(KwsSearch)) {
            switchSearch(KwsSearch);
        }
    }

    @Override
    public void onPartialResult(Hypothesis hypothesis) {
        if (hypothesis != null) {
            String result = hypothesis.getHypstr();
            if (result != null) {
                RelationActionClass relationActionClass = _relationshipHelper.GetRelationActionClass(result);
                try {
                    handleSpeechResult(relationActionClass);
                } catch (Exception exception) {
                    broadcastException(exception);
                }
            }
        }
    }

    @Override
    public void onResult(Hypothesis hypothesis) {
        if (hypothesis != null) {
            String result = hypothesis.getHypstr();
            if (result != null) {
                RelationActionClass relationActionClass = _relationshipHelper.GetRelationActionClass(result);
                try {
                    handleSpeechResult(relationActionClass);
                    return;
                } catch (Exception exception) {
                    broadcastException(exception);
                }
            }
        }
        broadcastException(new Exception("onResult received some null value!"));
    }

    @Override
    public void onError(Exception exception) {
        broadcastException(exception);
    }

    @Override
    public void onTimeout() {
        switchSearch(KwsSearch);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == PermissionRequestRecordAudio) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                _permissionRecordAudioGranted = true;
                // Recognizer initialization is a time-consuming and it involves IO, so we execute it in async task
                new SetupTask(this).execute();
            } else {
                _permissionRecordAudioGranted = false;
            }
            _broadcastController.SendBooleanBroadcast(BroadcastPermissionRecordAudioResult, BundlePermissionRecordAudioResult, _permissionRecordAudioGranted);
        }
    }

    private void broadcastException(Exception exception) {
        if (exception == null) {
            Logger.getInstance().Error(Tag, "broadcastException exception is null!");
            return;
        }

        Logger.getInstance().Error(Tag, exception.toString());

        if (_isInitialized && _broadcastController != null) {
            _broadcastController.SendStringBroadcast(BroadcastException, BundleException, exception.toString());
        }
    }

    private boolean checkRecordAudioPermission() {
        Logger.getInstance().Debug(Tag, "checkRecordAudioPermission");
        int permissionCheck = ContextCompat.checkSelfPermission(_context.getApplicationContext(), Manifest.permission.RECORD_AUDIO);
        if (permissionCheck != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions((Activity) _context, new String[]{Manifest.permission.RECORD_AUDIO}, PermissionRequestRecordAudio);
        }
        if (_isInitialized && _broadcastController != null) {
            _broadcastController.SendBooleanBroadcast(BroadcastPermissionRecordAudioResult, BundlePermissionRecordAudioResult, permissionCheck == PackageManager.PERMISSION_GRANTED);
        }
        return permissionCheck == PackageManager.PERMISSION_GRANTED;
    }

    private void setupRecognizer(File assetsDir) throws IOException {
        // The recognizer can be configured to perform multiple searches of different kind and switch between them
        _speechRecognizer = SpeechRecognizerSetup.defaultSetup()
                .setAcousticModel(new File(assetsDir, "en-us-ptm"))
                .setDictionary(new File(assetsDir, "cmudict-en-us.dict"))
                //.setRawLogDir(assetsDir) // To disable logging of raw audio comment out this call (takes a lot of space on the device)
                .getRecognizer();
        _speechRecognizer.addListener(this);

        /* In your application you might not need to add all those searches.
          They are added here for demonstration. You can leave just one. */

        // Create keyword-activation search.
        _speechRecognizer.addKeyphraseSearch(KwsSearch, KeyPhrase);
    }

    private void switchSearch(String searchValue) {
        if (searchValue == null) {
            return;
        }

        _speechRecognizer.stop();

        // If we are not spotting, start listening with timeout (10000 ms or 10 seconds).
        switch (searchValue) {
            case KwsSearch:
                _speechRecognizer.startListening(searchValue);
                break;
            default:
                _speechRecognizer.startListening(searchValue, 10000);
                break;
        }
    }

    private void handleSpeechResult(RelationActionClass relationActionClass) throws Exception {
        RelationActionClass.RelationAction relationAction = relationActionClass.GetRelationAction();
        ArrayList<String> resultParameter = relationActionClass.GetActionParameter();

        switch (relationAction) {
            case WirelessSocketOn:
                WirelessSocket wirelessSocketOn = _relationshipHelper.GetWirelessSocketByName(resultParameter, WirelessSocketService.getInstance().GetDataList());
                WirelessSocketService.getInstance().SetWirelessSocketState(wirelessSocketOn, true);
                break;
            case WirelessSocketOff:
                WirelessSocket wirelessSocketOff = _relationshipHelper.GetWirelessSocketByName(resultParameter, WirelessSocketService.getInstance().GetDataList());
                WirelessSocketService.getInstance().SetWirelessSocketState(wirelessSocketOff, false);
                break;
            case WirelessSocketToggle:
                WirelessSocket wirelessSocketToggle = _relationshipHelper.GetWirelessSocketByName(resultParameter, WirelessSocketService.getInstance().GetDataList());
                WirelessSocketService.getInstance().SetWirelessSocketState(wirelessSocketToggle, !wirelessSocketToggle.GetState());
                break;

            case WirelessSwitchToggle:
                WirelessSwitch wirelessSwitch = _relationshipHelper.GetWirelessSwitchByName(resultParameter, WirelessSwitchService.getInstance().GetDataList());
                WirelessSwitchService.getInstance().ToggleWirelessSwitch(wirelessSwitch);
                break;

            case PlayYoutube:
                _broadcastController.SendIntBroadcast(BroadcastAudioMedia, BundleAudioMedia, MediaState.PlayYoutube.ordinal());
                break;
            case PlayRadio:
                _broadcastController.SendIntBroadcast(BroadcastAudioMedia, BundleAudioMedia, MediaState.PlayRadio.ordinal());
                break;
            case Pause:
                _broadcastController.SendIntBroadcast(BroadcastAudioMedia, BundleAudioMedia, MediaState.Pause.ordinal());
                break;
            case Stop:
                _broadcastController.SendIntBroadcast(BroadcastAudioMedia, BundleAudioMedia, MediaState.Stop.ordinal());
                break;

            case WeatherCurrent:
                WeatherModel currentWeather = (WeatherModel) OpenWeatherService.getInstance().GetCurrentWeather();
                String ttsSpeakCurrentWeather = createTtsSpeak(currentWeather);
                _broadcastController.SendStringBroadcast(BroadcastTtsSpeak, BundleTtsSpeak, ttsSpeakCurrentWeather);
                break;

            case WeatherForecast:
                ForecastModel forecastWeather = (ForecastModel) OpenWeatherService.getInstance().GetForecastWeather();
                String ttsSpeakForecastWeather = createTtsSpeak(forecastWeather);
                _broadcastController.SendStringBroadcast(BroadcastTtsSpeak, BundleTtsSpeak, ttsSpeakForecastWeather);
                break;

            case GetLight:
                PuckJs puckJs = _relationshipHelper.GetPuckJsByArea(resultParameter, PuckJsService.getInstance().GetDataList());
                String ttsSpeakPuckJs = createTtsSpeak(puckJs);
                _broadcastController.SendStringBroadcast(BroadcastTtsSpeak, BundleTtsSpeak, ttsSpeakPuckJs);
                break;

            case GetTemperature:
                Temperature temperature = _relationshipHelper.GetTemperatureByArea(resultParameter, TemperatureService.getInstance().GetDataList());
                String ttsSpeakTemperature = createTtsSpeak(temperature);
                _broadcastController.SendStringBroadcast(BroadcastTtsSpeak, BundleTtsSpeak, ttsSpeakTemperature);
                break;

            case Unknown:
            default:
                broadcastException(new NullPointerException(String.format(Locale.getDefault(), "Unknown action for result %s", resultParameter)));
                break;
        }
    }

    private String createTtsSpeak(WeatherModel currentWeather) {
        if (currentWeather == null) {
            return "No current weather available";
        }

        return String.format(Locale.getDefault(), "Current weather in %s is %s. It has %.2f degree.", currentWeather.GetCity(), currentWeather.GetWeatherCondition().GetDescription(), currentWeather.GetTemperature());
    }

    private String createTtsSpeak(ForecastModel forecastWeather) {
        if (forecastWeather == null) {
            return "No forecast weather available";
        }

        NotificationContentDto notificationContent = NotificationContentConverter.GetForecastWeather(forecastWeather.GetList());
        return notificationContent.GetTitle() + notificationContent.GetText();
    }

    private String createTtsSpeak(PuckJs puckJs) {
        if (puckJs == null) {
            return "No valid puck found";
        }

        Room room = RoomService.getInstance().GetByUuid(puckJs.GetRoomUuid());
        return String.format(Locale.getDefault(), "There is a relative light value of %d in the %s", puckJs.GetLightValue(), room.GetName());
    }

    private String createTtsSpeak(Temperature temperature) {
        if (temperature == null) {
            return "No valid temperature found";
        }

        Room room = RoomService.getInstance().GetByUuid(temperature.GetRoomUuid());
        return String.format(Locale.getDefault(), "It has %.2f degree in the %s", temperature.GetTemperature(), room.GetName());
    }

    private static class SetupTask extends AsyncTask<Void, Void, Exception> {
        private WeakReference<VoiceRecognitionService> _voiceRecognitionServiceWeakReference;

        SetupTask(@NonNull VoiceRecognitionService voiceRecognitionService) {
            _voiceRecognitionServiceWeakReference = new WeakReference<>(voiceRecognitionService);
        }

        @Override
        protected Exception doInBackground(Void... voids) {
            try {
                Assets assets = new Assets(_voiceRecognitionServiceWeakReference.get()._context);
                File assetDir = assets.syncAssets();
                _voiceRecognitionServiceWeakReference.get().setupRecognizer(assetDir);
            } catch (IOException exception) {
                Logger.getInstance().Error(Tag, exception.toString());
                return exception;
            }
            return null;
        }

        @Override
        protected void onPostExecute(Exception result) {
            _voiceRecognitionServiceWeakReference.get()._broadcastController.SendBooleanBroadcast(BroadcastSetupSpeechRecognizerResult, BundleSetupSpeechRecognizerResult, result == null);
        }
    }
}

File: voicerecognition/src/main/java/guepardoapps/lucahome/voicerecognition/services/VoiceRecognitionService.kt
package guepardoapps.lucahome.voicerecognition.services

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.AsyncTask
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import edu.cmu.pocketsphinx.*
import guepardoapps.lib.openweather.enums.WeatherCondition
import guepardoapps.lib.openweather.extensions.*
import guepardoapps.lib.openweather.models.RxOptional
import guepardoapps.lib.openweather.models.WeatherCurrent
import guepardoapps.lib.openweather.models.WeatherForecast
import guepardoapps.lib.openweather.services.openweather.OpenWeatherService
import guepardoapps.lucahome.common.controller.TtsController
import guepardoapps.lucahome.common.extensions.common.doubleFormat
import guepardoapps.lucahome.common.models.puckjs.PuckJs
import guepardoapps.lucahome.common.models.temperature.Temperature
import guepardoapps.lucahome.common.services.position.PositionService
import guepardoapps.lucahome.common.services.puckjs.PuckJsService
import guepardoapps.lucahome.common.services.room.RoomService
import guepardoapps.lucahome.common.services.temperature.TemperatureService
import guepardoapps.lucahome.common.services.wirelesssocket.WirelessSocketService
import guepardoapps.lucahome.common.services.wirelessswitch.WirelessSwitchService
import guepardoapps.lucahome.common.utils.Logger
import guepardoapps.lucahome.voicerecognition.enums.Action
import guepardoapps.lucahome.voicerecognition.enums.InitializeResult
import guepardoapps.lucahome.voicerecognition.enums.MediaState
import guepardoapps.lucahome.voicerecognition.models.RelationAction
import guepardoapps.lucahome.voicerecognition.helper.*
import io.reactivex.schedulers.Schedulers
import java.io.File
import java.io.IOException
import java.lang.Exception
import java.lang.ref.WeakReference

class VoiceRecognitionService private constructor() : RecognitionListener, ActivityCompat.OnRequestPermissionsResultCallback, IVoiceRecognitionService {
    private val tag: String = VoiceRecognitionService::class.java.simpleName

    private val permissionRequestRecordAudio = 345243854

    private val kwsSearch = "wakeup"
    private val keyPhrase = "hey luca"

    private var isInitialized: Boolean = false
    private var permissionRecordAudioGranted: Boolean = false

    private var context: Context? = null
    private var speechRecognizer: SpeechRecognizer? = null
    private var ttsController: TtsController? = null

    private var onVoiceRecognitionService: OnVoiceRecognitionService? = null

    init {
    }

    private object Holder {
        @SuppressLint("StaticFieldLeak")
        val instance: VoiceRecognitionService = VoiceRecognitionService()
    }

    companion object {
        val instance: VoiceRecognitionService by lazy { Holder.instance }

        const val broadcastAudioMedia = "gueopardoapps.lucahome.voicerecognition.services.broadcast.audio_media"
        const val bundleAudioMedia = "bundleAudioMedia"
    }

    override fun initialize(context: Context, onVoiceRecognitionService: OnVoiceRecognitionService): InitializeResult {
        if (isInitialized) {
            Logger.instance.debug(tag, "Already initialized!")
            return InitializeResult.AlreadyInitialized
        }

        if (context !is Activity) {
            Logger.instance.debug(tag, "Context is not of type Activity!")
            return InitializeResult.ContextNotOfTypeActivity
        }

        Logger.instance.debug(tag, "Initialize")

        this.context = context
        this.onVoiceRecognitionService = onVoiceRecognitionService
        this.ttsController = TtsController()

        try {
            permissionRecordAudioGranted = checkRecordAudioPermission()
        } catch (exception: Exception) {
            Logger.instance.error(tag, exception)
            this.onVoiceRecognitionService?.onPermissionGranted(false)
        }

        if (permissionRecordAudioGranted) {
            SetupTask(this).execute()
        }

        isInitialized = true
        return InitializeResult.InitializeSuccess
    }

    override fun dispose() {
        Logger.instance.debug(tag, "Dispose")
        if (speechRecognizer != null) {
            speechRecognizer!!.cancel()
            speechRecognizer!!.shutdown()
        }
        isInitialized = false
    }

    override fun requestRecordAudioPermission(): Boolean {
        Logger.instance.verbose(tag, "RequestRecordAudioPermission")
        if (permissionRecordAudioGranted) {
            Logger.instance.verbose(tag, "Already granted!")
            return true
        }

        return try {
            permissionRecordAudioGranted = checkRecordAudioPermission()
            permissionRecordAudioGranted
        } catch (exception: Exception) {
            Logger.instance.error(tag, exception)
            false
        }
    }

    override fun onBeginningOfSpeech() {
    }

    override fun onEndOfSpeech() {
        if (speechRecognizer!!.searchName != kwsSearch) {
            switchSearch(kwsSearch)
        }
    }

    override fun onPartialResult(hypothesis: Hypothesis?) {
        if (hypothesis != null) {
            val result = hypothesis.hypstr
            if (result != null) {
                val relationActionClass = getRelationActionClass(result)
                try {
                    handleSpeechResult(relationActionClass)
                } catch (exception: Exception) {
                    Logger.instance.error(tag, exception)
                }
            }
        }
    }

    override fun onResult(hypothesis: Hypothesis?) {
        if (hypothesis != null) {
            val result = hypothesis.hypstr
            if (result != null) {
                val relationActionClass = getRelationActionClass(result)
                try {
                    handleSpeechResult(relationActionClass)
                    return
                } catch (exception: Exception) {
                    Logger.instance.error(tag, exception)
                }

            }
        }
    }

    override fun onError(exception: Exception?) {
        Logger.instance.error(tag, exception)
    }

    override fun onTimeout() {
        switchSearch(kwsSearch)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        if (requestCode == permissionRequestRecordAudio) {
            if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                permissionRecordAudioGranted = true
                // Recognizer initialization is a time-consuming and it involves IO, so we execute it in async task
                SetupTask(this).execute()
            } else {
                permissionRecordAudioGranted = false
            }
        }
        onVoiceRecognitionService?.onPermissionGranted(permissionRecordAudioGranted)
    }

    private fun checkRecordAudioPermission(): Boolean {
        Logger.instance.debug(tag, "checkRecordAudioPermission")
        val permissionCheck = ContextCompat.checkSelfPermission(context!!.applicationContext, Manifest.permission.RECORD_AUDIO)
        if (permissionCheck != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(context as Activity, arrayOf(Manifest.permission.RECORD_AUDIO), permissionRequestRecordAudio)
        }

        if (isInitialized) {
            onVoiceRecognitionService?.onPermissionGranted(permissionCheck == PackageManager.PERMISSION_GRANTED)
        }

        return permissionCheck == PackageManager.PERMISSION_GRANTED
    }

    private fun switchSearch(searchValue: String?) {
        if (searchValue == null) {
            return
        }
        speechRecognizer?.stop()
        // If we are not spotting, start listening with timeout (10000 ms or 10 seconds).
        when (searchValue) {
            kwsSearch -> speechRecognizer?.startListening(searchValue)
            else -> speechRecognizer?.startListening(searchValue, 10000)
        }
    }

    @Throws(IOException::class)
    private fun setupRecognizer(assetsDir: File) {
        // The recognizer can be configured to perform multiple searches of different kind and switch between them
        speechRecognizer = SpeechRecognizerSetup.defaultSetup()
                .setAcousticModel(File(assetsDir, "en-us-ptm"))
                .setDictionary(File(assetsDir, "cmudict-en-us.dict"))
                //.setRawLogDir(assetsDir) // To disable logging of raw audio comment out this call (takes a lot of space on the device)
                .recognizer
        speechRecognizer?.addListener(this)

        /* In your application you might not need to add all those searches.
          They are added here for demonstration. You can leave just one. */

        // Create keyword-activation search.
        speechRecognizer?.addKeyphraseSearch(kwsSearch, keyPhrase)
    }

    @Throws(Exception::class)
    private fun handleSpeechResult(relationAction: RelationAction) {
        val action = relationAction.action
        val parameter = relationAction.parameter

        when (action) {
            Action.WirelessSocketOn -> {
                val wirelessSocketOn = getWirelessSocketByName(parameter, WirelessSocketService.instance.get())
                WirelessSocketService.instance.setState(wirelessSocketOn!!, true)
            }
            Action.WirelessSocketOff -> {
                val wirelessSocketOff = getWirelessSocketByName(parameter, WirelessSocketService.instance.get())
                WirelessSocketService.instance.setState(wirelessSocketOff!!, false)
            }
            Action.WirelessSocketToggle -> {
                val wirelessSocketToggle = getWirelessSocketByName(parameter, WirelessSocketService.instance.get())
                WirelessSocketService.instance.setState(wirelessSocketToggle!!, !wirelessSocketToggle.state)
            }

            Action.WirelessSwitchToggle -> {
                val wirelessSwitch = getWirelessSwitchByName(parameter, WirelessSwitchService.instance.get())
                WirelessSwitchService.instance.toggle(wirelessSwitch!!)
            }

            Action.PlayYoutube -> {
                val bundle = Bundle()
                bundle.putInt(bundleAudioMedia, MediaState.PlayYoutube.ordinal)
                val intent = Intent(broadcastAudioMedia)
                intent.putExtras(bundle)
                context?.sendBroadcast(intent)
            }
            Action.PlayRadio -> {
                val bundle = Bundle()
                bundle.putInt(bundleAudioMedia, MediaState.PlayRadio.ordinal)
                val intent = Intent(broadcastAudioMedia)
                intent.putExtras(bundle)
                context?.sendBroadcast(intent)
            }
            Action.Pause -> {
                val bundle = Bundle()
                bundle.putInt(bundleAudioMedia, MediaState.Pause.ordinal)
                val intent = Intent(broadcastAudioMedia)
                intent.putExtras(bundle)
                context?.sendBroadcast(intent)
            }
            Action.Stop -> {
                val bundle = Bundle()
                bundle.putInt(bundleAudioMedia, MediaState.Stop.ordinal)
                val intent = Intent(broadcastAudioMedia)
                intent.putExtras(bundle)
                context?.sendBroadcast(intent)
            }

            Action.WeatherCurrent -> {
                OpenWeatherService.instance.weatherCurrentPublishSubject
                        .subscribeOn(Schedulers.io())
                        .first(RxOptional(WeatherCurrent()))
                        .subscribe(
                                { response ->
                                    Logger.instance.verbose(tag, "Received weather current in subscribe!")

                                    if (response != null) {
                                        val value = response.value as WeatherCurrent
                                        if (value.weatherCondition != WeatherCondition.Null) {
                                            this.ttsController?.speak(createTtsSpeak(value))
                                        } else {
                                            this.ttsController?.speak("Sorry, no current weather data")
                                        }
                                    } else {
                                        this.ttsController?.speak("Sorry, no current weather data")
                                    }
                                },
                                { responseError ->
                                    Logger.instance.error(tag, responseError)
                                    this.ttsController?.speak("Sorry, no current weather data")
                                })
            }

            Action.WeatherForecast -> {
                OpenWeatherService.instance.weatherForecastPublishSubject
                        .subscribeOn(Schedulers.io())
                        .first(RxOptional(WeatherForecast()))
                        .subscribe(
                                { response ->
                                    Logger.instance.verbose(tag, "Received weather forecast in subscribe!")

                                    if (response != null) {
                                        val value = response.value as WeatherForecast
                                        if (value.list.isNotEmpty()) {
                                            this.ttsController?.speak(createTtsSpeak(value))
                                        } else {
                                            this.ttsController?.speak("Sorry, no forecast weather data")
                                        }
                                    } else {
                                        this.ttsController?.speak("Sorry, no forecast weather data")
                                    }
                                },
                                { responseError ->
                                    Logger.instance.error(tag, responseError)
                                    this.ttsController?.speak("Sorry, no forecast weather data")
                                })
            }

            Action.GetLight -> {
                val puckJs = getPuckJsByArea(parameter, PuckJsService.instance.get())
                this.ttsController?.speak(createTtsSpeak(puckJs))
            }

            Action.GetTemperature -> {
                val temperature = getTemperatureByArea(parameter, TemperatureService.instance.get())
                this.ttsController?.speak(createTtsSpeak(temperature))
            }

            Action.Unknown -> {
                val exception = NullPointerException("Unknown action for result $parameter")
                Logger.instance.error(tag, exception)
                onVoiceRecognitionService?.onError(exception)
            }
        }
    }

    private fun createTtsSpeak(weatherCurrent: WeatherCurrent?): String {
        if (weatherCurrent == null) {
            return "No current weather available"
        }
        return "Current weather in ${weatherCurrent.city} is ${weatherCurrent.weatherCondition.description}. It has ${weatherCurrent.temperature.doubleFormat(2)} degree."
    }

    private fun createTtsSpeak(weatherForecast: WeatherForecast?): String {
        if (weatherForecast == null) {
            return "No forecast weather available"
        }
        return "It will be " + weatherForecast.getMostWeatherCondition().description + "." +
                "Temperature will be from " + weatherForecast.getMinTemperature().doubleFormat(1) + "${0x00B0.toChar()}C to " + weatherForecast.getMaxTemperature().doubleFormat(1) + "${0x00B0.toChar()}C." +
                "Air pressure will be from " + weatherForecast.getMinPressure().doubleFormat(1) + "mBar to " + weatherForecast.getMaxPressure().doubleFormat(1) + "mBar." +
                "Humidity will be from " + weatherForecast.getMinHumidity().doubleFormat(1) + "% to " + weatherForecast.getMaxHumidity().doubleFormat(1) + "%"
    }

    private fun createTtsSpeak(puckJs: PuckJs?): String {
        if (puckJs == null) {
            return "No puck js found"
        }
        val position = PositionService.instance.currentPosition
        val room = RoomService.instance.get(puckJs.roomUuid)
        return "There is a relative light value of ${position.lightValue.doubleFormat(2)} in the ${room?.name}"
    }

    private fun createTtsSpeak(temperature: Temperature?): String {
        if (temperature == null) {
            return "No temperature found"
        }
        return "It has ${temperature.value.doubleFormat(2)} degree in the ${temperature.area}"
    }

    private class SetupTask internal constructor(voiceRecognitionService: VoiceRecognitionService) : AsyncTask<Void, Void, Exception>() {
        private val tag: String = SetupTask::class.java.simpleName

        private val _voiceRecognitionServiceWeakReference: WeakReference<VoiceRecognitionService> = WeakReference(voiceRecognitionService)

        override fun doInBackground(vararg voids: Void): Exception? {
            try {
                val assets = Assets(_voiceRecognitionServiceWeakReference.get()!!.context)
                val assetDir = assets.syncAssets()
                _voiceRecognitionServiceWeakReference.get()!!.setupRecognizer(assetDir)
            } catch (exception: IOException) {
                Logger.instance.error(tag, exception.toString())
                return exception
            }

            return null
        }

        override fun onPostExecute(result: Exception?) {
            _voiceRecognitionServiceWeakReference.get()!!.onVoiceRecognitionService?.onInitializationFinished(result == null)
        }
    }
}